server/src/main/java/org/elasticsearch/action/support/RetryableAction.java

```diff
@@ -0,0 +1,176 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRunnable;
+import org.elasticsearch.common.Randomness;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayDeque;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A action that will be retried on failure if {@link RetryableAction#shouldRetry(Exception)} returns true.
+ * The executor the action will be executed on can be defined in the constructor. Otherwise, SAME is the
+ * default. The action will be retried with exponentially increasing delay periods until the timeout period
+ * has been reached.
+ */
+public abstract class RetryableAction<Response> {
+
+    private final Logger logger;
+
+    private final AtomicBoolean isDone = new AtomicBoolean(false);
+    private final ThreadPool threadPool;
+    private final long initialDelayMillis;
+    private final long timeoutMillis;
+    private final long startMillis;
+    private final ActionListener<Response> finalListener;
+    private final String retryExecutor;
+
+    public RetryableAction(Logger logger, ThreadPool threadPool, TimeValue initialDelay, TimeValue timeoutValue,
+                           ActionListener<Response> listener) {
+        this(logger, threadPool, initialDelay, timeoutValue, listener, ThreadPool.Names.SAME);
+    }
+
+    public RetryableAction(Logger logger, ThreadPool threadPool, TimeValue initialDelay, TimeValue timeoutValue,
+                           ActionListener<Response> listener, String executor) {
+        this.logger = logger;
+        this.threadPool = threadPool;
+        this.initialDelayMillis = initialDelay.getMillis();
+        if (initialDelayMillis < 1) {
+            throw new IllegalArgumentException("Initial delay was less than 1 millisecond: " + initialDelay);
+        }
+        this.timeoutMillis = Math.max(timeoutValue.getMillis(), 1);
+        this.startMillis = threadPool.relativeTimeInMillis();
+        this.finalListener = listener;
+        this.retryExecutor = executor;
+    }
+
+    public void run() {
+        final RetryingListener retryingListener = new RetryingListener(initialDelayMillis, null);
+        final Runnable runnable = createRunnable(retryingListener);
+        threadPool.executor(retryExecutor).execute(runnable);
```

Comment:
I renamed executor and made it an executor. I felt weird about the fact that the first attempt might use a different executor than the retries. Although it does not really matter for this usage since this one always uses SAME.